### PR TITLE
[WIP] Add Devfile Parser for Version 2.0.0

### LINKED ISSUE #2918: Devfile 2.0 parser
/kind user-story

## User Story
Add devfile parser support to read devfiles that users the devfile 2.0 specification. Details of devfile 2.0 specification can be found under:
https://github.com/che-incubator/devworkspace-api/issues/15

Devfile schema link: TBD

## Acceptance Criteria 
- [x] It should detect the version of devfile based on the apiVersion in devfile to decide if devfile 1.0 parser or devfile 2.0 parser is being used
- [x] It should generate the data model that represents devfile 2.0 model

## Links 
- Related issue: 

/kind user-story
/area devfile

### PR DESCRIPTION
**Pull request by adisky**
Signed-off-by: adisky <adsharma@redhat.com>

**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
Added Devfile V2 Go Structures
Added converter for v2 to common types

**Which issue(s) this PR fixes**:

Fixes #2918 

**How to test changes / Special notes to the reviewer**:
create, odo component from example devfileV2.yaml, https://github.com/openshift/odo/blob/700248d328903375489610f4471999daef817d73/tests/examples/source/devfiles/nodejs/devfileV2.yaml 
It should get parsed and deployed sucessfully.

### CODE REVIEW

File: pkg/devfile/parser/data/2.0.0/components.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	for _, v := range c.Env {
+		envs = append(envs, convertV2EnvToCommon(*v))
+	}
+
+	var volumes []common.DockerimageVolume
+	for _, v := range c.VolumeMounts {
+		volumes = append(volumes, convertV2VolumeToCommon(*v))
+	}
+
+	// TODO: Args, Command not converted (as no matching types found)
```
Comment (girishramnani) [inline pkg/devfile/parser/data/2.0.0/components.go:118]: TODO?

File: tests/examples/source/devfiles/nodejs/devfileV2.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      mountSources: true
+      name: "runtime"
+      memoryLimit: 1024Mi
+      env:
+      - name: FOO
+        value: "bar"
+      endpoints:
+        - name: '3000/tcp'
+          targetPort: 3000
+          # Not parsing currently, added to validate JSON Schema
```
Comment (girishramnani) [inline tests/examples/source/devfiles/nodejs/devfileV2.yaml:20]: why are we not parsing `configuration`?

Comment (adisky) [reply tests/examples/source/devfiles/nodejs/devfileV2.yaml:20]: Need to update the cooment, We are parsing but not using the values.

**Comment by kadel** [comment]
some examples of devfile v2 https://github.com/sbose78/stack-registry/tree/sbose78-patch-1/stacks

File: pkg/devfile/parser/data/1.0.0/components.go [hunk 3]
```diff
@@ -1,21 +1,37 @@
 package version100
 
 import (
-	"strings"
-
 	"github.com/openshift/odo/pkg/devfile/parser/data/common"
 )
 
-// GetComponents returns the slice of DevfileComponent objects parsed from the Devfile
+func (d *Devfile100) GetMetadata() common.DevfileMetadata {
```
Comment (prietyc123) [inline pkg/devfile/parser/data/1.0.0/components.go:7]: Shouldn't we have comment for the function `GetMetadata` like what it does?

File: pkg/devfile/parser/data/1.0.0/components.go [hunk 4]
```diff
@@ -24,17 +40,157 @@ func (d *Devfile100) GetAliasedComponents() []common.DevfileComponent {
...
+
+		if *action.Type == DevfileCommandTypeExec {
+			exec = common.Exec{
+				Attributes:  c.Attributes,
+				CommandLine: *action.Command,
+				Component:   *action.Component,
+				Group:       getGroup(c.Name),
+				Id:          c.Name,
+				WorkingDir:  action.Workdir,
+				// Env:
```
Comment (prietyc123) [inline pkg/devfile/parser/data/1.0.0/components.go:91]: Extra commented fields

File: pkg/devfile/parser/context/apiVersion.go [hunk 5]
```diff
@@ -19,19 +19,34 @@ func (d *DevfileCtx) SetDevfileAPIVersion() error {
...
-	}
+	var apiVer string
+
+	// Get "apiVersion" value from map for devfile V1
+	apiVersion, okApi := r["apiVersion"]
+
+	// Get "schemaVersion" value from map for devfile V2
+	schemaVersion, okSchema := r["schemaVersion"]
+
+	if okApi {
```
Comment (prietyc123) [inline pkg/devfile/parser/context/apiVersion.go:30]: Could you please reduce this nested if else?

**Comment by adisky** [comment]
closing this PR as development for v2, parser and adaptor refactor is going in this branch https://github.com/openshift/odo/tree/parser-v2
/close

**Pull request closed without merging**